### PR TITLE
add cancel task option to all AMA task models

### PR DESCRIPTION
**Pull request by joeyyang**
Resolves https://github.com/department-of-veterans-affairs/caseflow/issues/9310

### Description
Adds the option to cancel task to all AMA task models (`judge_task.rb`, `generic_task.rb`, etc., excludes `attorney_task.rb` and `{any}_legacy_task.rb`)

Example:
screenshot 2019-02-19 11 33 40

### CODE REVIEW

File: app/models/tasks/generic_task.rb [hunk 1]
```diff
@@ -24,22 +24,24 @@ def available_actions(user)
       return [
         Constants.TASK_ACTIONS.ASSIGN_TO_TEAM.to_h,
         Constants.TASK_ACTIONS.REASSIGN_TO_PERSON.to_h,
-        Constants.TASK_ACTIONS.CANCEL_TASK.to_h,
-        Constants.TASK_ACTIONS.MARK_COMPLETE.to_h
+        Constants.TASK_ACTIONS.MARK_COMPLETE.to_h,
+        Constants.TASK_ACTIONS.CANCEL_TASK.to_h
```
Comment (joeyyang) [inline app/models/tasks/generic_task.rb]: make `CANCEL_TASK` go last everywhere it shows up

File: app/models/tasks/judge_task.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 class JudgeTask < Task
   def available_actions(user)
-    additional_available_actions(user).unshift(Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h)
+    additional_available_actions(user).unshift(Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h).push(Constants.TASK_ACTIONS.CANCEL_TASK.to_h)
```
Comment (joeyyang) [inline app/models/tasks/judge_task.rb]: there's gotta be a better way to do this

Comment (lowellrex) [reply app/models/tasks/judge_task.rb]: I think this is fine. Though I guess we could do something like the following if you prefer:

```ruby
[
  Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h,
  additional_available_actions(user),
  Constants.TASK_ACTIONS.CANCEL_TASK.to_h
].flatten
```

Comment (joeyyang) [reply app/models/tasks/judge_task.rb]: i like this much better!

File: app/models/tasks/judge_task.rb [hunk 3]
```diff
@@ -1,6 +1,10 @@
 class JudgeTask < Task
   def available_actions(user)
-    additional_available_actions(user).unshift(Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h)
+    [
+      Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h,
+      additional_available_actions(user),
+      Constants.TASK_ACTIONS.CANCEL_TASK.to_h
```
Comment (joeyyang) [inline app/models/tasks/judge_task.rb]: ask lauren

File: app/models/tasks/schedule_hearing_task.rb [hunk 4]
```diff
@@ -110,7 +110,8 @@ def available_actions(user)
       return [
         Constants.TASK_ACTIONS.SCHEDULE_VETERAN.to_h,
         Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h,
-        Constants.TASK_ACTIONS.WITHDRAW_HEARING.to_h
+        Constants.TASK_ACTIONS.WITHDRAW_HEARING.to_h,
```
Comment (joeyyang) [inline app/models/tasks/schedule_hearing_task.rb]: removing cancel option here, since "Withdraw hearing" also has `"value": "modal/cancel_task"`, which fails `app/models/task.rb#available_actions_unwrapper:58`

File: app/models/tasks/colocated_task.rb [hunk 5]
```diff
@@ -45,7 +45,8 @@ def available_actions(user)
 
     available_actions_with_conditions([
                                         Constants.TASK_ACTIONS.PLACE_HOLD.to_h,
-                                        Constants.TASK_ACTIONS.ASSIGN_TO_PRIVACY_TEAM.to_h
+                                        Constants.TASK_ACTIONS.ASSIGN_TO_PRIVACY_TEAM.to_h,
+                                        Constants.TASK_ACTIONS.CANCEL_TASK.to_h
```
Comment (lowellrex) [inline app/models/tasks/colocated_task.rb]: I think we need to update `ColocatedTask.update_location_in_vacols` to allow the cancellation of tasks to return the case to the correct location in VACOLS (as opposed to just the completion of tasks, as it is written now).

**Comment by lowellrex** [comment]
Looks good for the most part! Just a few thoughts:

1. Let's confirm with @laurjpeterson that we are alright to delay adding the ability to cancel `AttorneyTask`s.
2. Let's determine what it means for a `JudgeTask` to be cancelled. What happens if a judge cancels a `JudgeAssignTask`, should it return to the basket of cases to be distributed? Should we allow them to do this?
3. Let's update `ColocatedTask` to make sure we charge cancelled tasks associated with `LegacyAppeal`s back to the appropriate location in VACOLS.
4. Let's add tests for each of these scenarios to confirm that the states of parent tasks change as expected when we cancel child tasks.

**Comment by joeyyang** [comment]
@laurjpeterson should a judge be able to cancel any kind of `JudgeTask` (types are `JudgeQualityReviewTask`, `JudgeAssignTask`, `JudgeDecisionReviewTask`)?  @lowellrex and I thought that `JudgeQualityReviewTask`s were the only ones that could be cancelled in any meaningful way.

**Comment by joeyyang** [comment]
WIP: logic for if a task is cancelled?
- for a `ColocatedTask`, it should go back to the attorney who assigned it
  - for a `LegacyAppeal`, it should go back to whoever assigned it (look in VACOLS)
- for a `QualityReviewTask`, it should create a `BvaDispatchTask`? (should it be cancellable?)
- for a `GenericTask`, 
- for a `HearingAdminActionTask`
- for a `InformalHearingPresentationTask`
- for a `JudgeQualityReviewTask`

**Comment by laurjpeterson** [comment]
> 1. Let's confirm with @laurjpeterson that we are alright to delay adding the ability to cancel `AttorneyTask`s.
- Sure. Judges can just reassign attorney cases if they need to be taken off attorneys' plates.

> @laurjpeterson should a judge be able to cancel any kind of `JudgeTask` (types are `JudgeQualityReviewTask`, `JudgeAssignTask`, `JudgeDecisionReviewTask`)?  
> @lowellrex and I thought that `JudgeQualityReviewTask`s were the only ones that could be cancelled in any meaningful way and would like to propose that only `JudgeQualityReviewTask`s be cancellable.
- Def can't cancel a JudgeAssignTask, because it was assigned to the judge in docket order so it can't go backwards
- I agree that JudgeQR is cancellable
- I agree that shouldn't allow cancelling JudgeDecisionReview task. I might be persuaded otherwise, because I could see cases get assigned to a judge, but the judge goes on leave or travel board, and therefore can't actually sign that case. so they want to cancel it so their attorney can assign it to someone else. other ways to accomplish this are for judges to reassign cases to other judges (though I think this fails right now - https://github.com/department-of-veterans-affairs/caseflow/issues/7612) or for certain caseflow users to be able to reassign cases to other judges

File: app/models/tasks/colocated_task.rb [hunk 6]
```diff
@@ -88,10 +89,12 @@ def create_and_auto_assign_child_task(_options = {})
...
-    if saved_change_to_status? &&
-       completed? &&
-       appeal_type == LegacyAppeal.name &&
-       all_tasks_completed_for_appeal?
+    is_closed_legacy_appeal = saved_change_to_status? &&
+                              !active? &&
+                              appeal_type == LegacyAppeal.name &&
+                              all_tasks_closed_for_appeal?
+
+    if is_closed_legacy_appeal
```
Comment (lowellrex) [inline app/models/tasks/colocated_task.rb]: Can we rename this variable to something like `all_colocated_tasks_for_legacy_appeal_complete` or maybe even go back to not assigning the value of those conditions to a variable at all so we can sidestep the issue of precise variable naming in this case?

File: app/models/tasks/attorney_task.rb [hunk 7]
```diff
@@ -24,7 +24,10 @@ def available_actions(user)
                             else
                               Constants.TASK_ACTIONS.REVIEW_LEGACY_DECISION.to_h
                             end
-    [review_decision_label, Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h]
+    [
+      review_decision_label,
+      Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h
+    ]
```
Comment (lowellrex) [inline app/models/tasks/attorney_task.rb]: I think we can undo this change for a tidier commit if we'd like.

File: app/models/tasks/judge_task.rb [hunk 8]
```diff
@@ -4,7 +4,10 @@
 
 class JudgeTask < Task
   def available_actions(user)
-    additional_available_actions(user).unshift(Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h)
+    [
+      Constants.TASK_ACTIONS.ADD_ADMIN_ACTION.to_h,
+      additional_available_actions(user)
+    ].flatten
```
Comment (lowellrex) [inline app/models/tasks/judge_task.rb]: I think we can undo this change for a tidier commit if we'd like.

File: app/models/tasks/attorney_task.rb (hunk 7 above)
Comment (joeyyang) [reply app/models/tasks/attorney_task.rb]: I think showing each set of actions on its own line is much more readable, if that's ok!
